### PR TITLE
Nullable Annotations

### PR DESCRIPTION
**Pull request by gfs**
Enable Nullable and add nullable annotations on all OSS Gadget projects.

These changes required I make some design decisions about which things will/will not ever return/accept null.  

Discussion on those design decisions is welcome.

### CODE REVIEW

File: src/Shared/PackageManagers/GitHubProjectManager.cs [hunk 1]
```diff
@@ -76,14 +76,14 @@ public override async Task<IEnumerable<string>> DownloadVersion(PackageURL purl,
...
-                var url = $"{ENV_GITHUB_ENDPOINT}/{purl.Namespace}/{purl.Name}";
+                var url = $"{ENV_GITHUB_ENDPOINT}/{purl?.Namespace}/{purl?.Name}";
                 var invalidChars = Path.GetInvalidFileNameChars();
 
                 // TODO: Externalize this normalization
-                var fsNamespace = new String(purl.Namespace.Select(ch => invalidChars.Contains(ch) ? '_' : ch).ToArray());
-                var fsName = new String(purl.Name.Select(ch => invalidChars.Contains(ch) ? '_' : ch).ToArray());
-                var fsVersion = new String(purl.Version.Select(ch => invalidChars.Contains(ch) ? '_' : ch).ToArray());
-                var workingDirectory = string.IsNullOrWhiteSpace(purl.Version) ?
+                var fsNamespace = new string((purl?.Namespace?.Select(ch => invalidChars.Contains(ch) ? '_' : ch) ?? Array.Empty<char>()).ToArray());
```
Comment (jacobmsft) [inline src/Shared/PackageManagers/GitHubProjectManager.cs:83]: This can lead to a very different result if purl is null, and cause no exceptions, correct?

File: src/Shared/PackageManagers/GitHubProjectManager.cs [hunk 2]
```diff
@@ -177,7 +177,7 @@ public static IEnumerable<PackageURL> ExtractGitHubUris(PackageURL purl, string
 
             try
             {
-                foreach (Match match in GithubExtractorRegex.Matches(searchText))
+                foreach (Match match in GithubExtractorRegex.Matches(searchText).Where(match => match != null))
```
Comment (jacobmsft) [inline src/Shared/PackageManagers/GitHubProjectManager.cs:180]: Is there are reason match could be null?

File: src/Shared/PackageManagers/BaseProjectManager.cs [hunk 3]
```diff
@@ -127,7 +127,7 @@ public static async Task<JsonDocument> GetJsonCache(string uri, bool useCache =
                     }
                 }
             }
-
+            if (WebClient == null) { throw new NullReferenceException(nameof(WebClient)); }
```
[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/PackageManagers/BaseProjectManager.cs:130]: The point of nullable annotations is to make null reference accesses explicit. If I don't check here we get a warning on the next line because WebClient can be null before you try to GetAsync it so you need to handle it there anyway.

File: src/Shared/PackageManagers/BaseProjectManager.cs [hunk 4]
```diff
@@ -217,7 +213,7 @@ public static IEnumerable<PackageURL> ExtractGitHubPackageURLs(string content)
             // @TODO: Check the regex below; does this match GitHub's scheme?
             var githubRegex = new Regex(@"github\.com/([a-z0-9\-_\.]+)/([a-z0-9\-_\.]+)",
                                         RegexOptions.IgnoreCase);
-            foreach (Match match in githubRegex.Matches(content))
+            foreach (Match match in githubRegex.Matches(content).Where(match => match != null))
```
[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/PackageManagers/BaseProjectManager.cs:216]: Yes. According to intellisense.

File: src/Shared/PackageManagers/CargoProjectManager.cs [hunk 5]
```diff
@@ -51,7 +51,7 @@ public override async Task<IEnumerable<string>> DownloadVersion(PackageURL purl,
             var url = $"{ENV_CARGO_ENDPOINT}/api/v1/crates/{packageName}/{packageVersion}/download";
             try
             {
-                string targetName = $"cargo-{purl.ToStringFilename()}";
+                string targetName = $"cargo-{purl?.ToStringFilename()}";
```
[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/PackageManagers/CargoProjectManager.cs:54]: That's fine.  You want to just return null from DownloadVersion if Purl is null in all cases?

I'll need to make it consistent in all of the download manager implementations.

File: src/Shared/PackageManagers/GitHubProjectManager.cs (hunk 1 above)
Comment (gfs) [reply src/Shared/PackageManagers/GitHubProjectManager.cs:83]: If Purl is null it will set `fsNamespace` to an empty string.

Yes, this means this won't throw exceptions here if Purl is null.

File: src/Shared/PackageManagers/GitHubProjectManager.cs (hunk 2 above)
Comment (gfs) [reply src/Shared/PackageManagers/GitHubProjectManager.cs:180]: Intellisense has deemed it thus.

File: src/Shared/PackageManagers/CargoProjectManager.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/PackageManagers/CargoProjectManager.cs:54]: If you check the lines above this we first check that `purl?.Name` and `purl?.Version` are not equal to null and return an empty list.  We can do the same if `purl?.ToStringFileName()` is null.  Does that work?

```csharp
var packageName = purl?.Name;
var packageVersion = purl?.Version;
var downloadedPaths = new List<string>();

if (string.IsNullOrWhiteSpace(packageName) || string.IsNullOrWhiteSpace(packageVersion))
{
    Logger.Error("Unable to download [{0} {1}]. Both must be defined.", packageName, packageVersion);
    return downloadedPaths;
}

var url = $"{ENV_CARGO_ENDPOINT}/api/v1/crates/{packageName}/{packageVersion}/download";
try
{
    string targetName = $"cargo-{purl?.ToStringFilename()}";
```

File: src/Shared/Helpers/PackageDownloader.cs [hunk 6]
```diff
@@ -212,30 +212,32 @@ public async void ClearPackageLocalCopy()
...
                     }
-                    File.WriteAllText(outputFilename, metadata);
-                    downloadPaths.Add(outputFilename);
                 }
-            }
-            else
-            {
-                // only version download requests reach here
-                downloadPaths.AddRange(await this.packageManager.DownloadVersion(purl, doExtract, this.actualCaching));
-                this.downloadPaths = downloadPaths;
```
[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/Helpers/PackageDownloader.cs:238]: I see why that was being set now.

I think we need to do this a different way.  If would call this function twice you'd lose the first download paths.  Should it be a List instead?

[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/Helpers/PackageDownloader.cs:238]: I see we were already using a list.  I switched to using AddRange instead of =, so extra calls won't overwrite the download paths that already existed and added a Clear in the ClearLocal copies function.

File: src/Shared/PackageManagers/BaseProjectManager.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/PackageManagers/BaseProjectManager.cs:130]: Finally,
1. This is not a server. 
2. A case where we *are* throwing an exception, not silently eating it.  Its up to the consumer to handle it properly, but we can't do anything about WebClient being null.  If you want to throw this null reference exception in the constructor instead when you instantiate webclient and not allow it to be nullable I can make that change.  
3. I don't think this is a boneheaded exception or close to the example they're talking about.  We can't Environment.Exit here the closest thing we can do is throw an exception.  This is as "fail hard" as you can possibly get in this case, so we are already following their advice.

Using Nullable annotations you must handle every case of nulls to generally *avoid* NullReferenceException, though in this case we don't have any choice but to throw a NullReferenceException.

[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/PackageManagers/BaseProjectManager.cs]: I switched to handling this by throwing a NullReferenceException in the constructor instead.

File: src/Shared/PackageManagers/GitHubProjectManager.cs (hunk 1 above)
Comment (gfs) [reply src/Shared/PackageManagers/GitHubProjectManager.cs]: If we should throw exceptions here we need to handle them explicitly.  I can restore the old "throw an exception" behavior if that's desired.

Comment (jacobmsft) [reply src/Shared/PackageManagers/GitHubProjectManager.cs]: yes, I think we need to do so...if we have a logic where it shouldn't proceed without a valid value, we should call object.hasValue() (??) and return failure/throw exception

Comment (gfs) [reply src/Shared/PackageManagers/GitHubProjectManager.cs]: This is fixed now. Check the latest.

File: src/oss-characteristics/CharacteristicTool.cs [hunk 7]
```diff
@@ -46,46 +46,48 @@ public class CharacteristicTool
         static void Main(string[] args)
         {
             var characteristicTool = new CharacteristicTool();
-            Logger.Debug($"Microsoft OSS Gadget - {TOOL_NAME} {VERSION}");
+            Logger?.Debug($"Microsoft OSS Gadget - {TOOL_NAME} {VERSION}");
```
Comment (jacobmsft) [inline src/oss-characteristics/CharacteristicTool.cs:49]: Do we need to do ?. on Logger? If logger is null, then pretty much all error info is unavailable and we need to know about it

Comment (gfs) [reply src/oss-characteristics/CharacteristicTool.cs:49]: Unless you want to crash, yes.

File: src/Shared/Helpers/OutputBuilder.cs [hunk 8]
```diff
@@ -31,33 +31,34 @@ enum OutputFormat
...
-        static string _Version = null;
-        string Version
+        static string? _Version = null;
+        string? Version
         {
             get
             {
-                _Version ??= Assembly.GetEntryAssembly().
-                    GetCustomAttribute<AssemblyFileVersionAttribute>().Version.ToString();
+                _Version ??= Assembly.GetEntryAssembly()?.
```
[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/Helpers/OutputBuilder.cs:49]: Sure. I can add a default value so Version isn't nullable.

[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/Helpers/OutputBuilder.cs:49]: I disagree that you need to crash though. You can run just fine without knowing the app version.

[the remark this answers is not in the document]
Comment (jacobmsft) [reply src/Shared/Helpers/OutputBuilder.cs:49]: Its not just running the app though, the sarif location property fields are also affected by this.

[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/Helpers/OutputBuilder.cs:49]: Can you comment on the line you're referring to?

File: src/oss-characteristics/CharacteristicTool.cs (hunk 7 above)
Comment (jacobmsft) [reply src/oss-characteristics/CharacteristicTool.cs:49]: Yes, so we check that the Logger is not null during Logger creation, and then assume its not null everywhere else, doing Logger.Warn instead of Logger?.Warn. There is no reason for it to be null anywhere else.
Implicit crashes are not a bad thing in case of catastrophic failures like Out of memory or non working terminal/logging system - since we don't have any way to resolve it, it might be better to crash and let the user know. Wont be practically possible to handle all of them anyways.

Comment (gfs) [reply src/oss-characteristics/CharacteristicTool.cs:49]: I'll make logger non-nullable.

File: src/Shared/Helpers/OutputBuilder.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jacobmsft) [reply src/Shared/Helpers/OutputBuilder.cs:49]: line 147 and 148 is where these fields are used. But we dont have anything to do about this other than set default values

[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/Helpers/OutputBuilder.cs:49]: I'll add default value.

**Comment by jacobmsft** [comment]
Need to modify the tests to add back initialization, failing because of null WebClient
